src/morphological_operations.jl

```diff
@@ -99,3 +99,90 @@ which is defined as the arithmetic difference between the internal and the exter
 `region` allows you to control the dimensions over which this operation is performed.
 """
 morpholaplace(img::AbstractArray, region=coords_spatial(img)) = dilate(img, region) + erode(img, region) - 2img
+
+
+struct FalsePadded
+    img::AbstractArray{Bool,2}
+end
+import Base.getindex
+function getindex(img::FalsePadded, r::Integer, c::Integer)
+    if (r==0) || (c==0)
+        return false
+    elseif (r> size(img.img,1)) || (c>size(img.img,2))
+        return false
+   else
+        return img.img[r,c]
+   end
+end
+
+"""
+```
+function thinning_iteration(img::AbstractArray{Bool,2}; odd_iteration::Bool)
+```
+The thining iteration evaluates three conditions in order to determine which pixels of the image should be removed from img.
+The three conditions are explained in:
+Guo, Z., & Hall, R. W. (1989). Parallel thinning with two-subiteration algorithms. Communications of the ACM, 32(3), 359-373.
+"""
+function thinning_iteration!(img_ori::AbstractArray{Bool,2}, odd_iteration::Bool)
+    local marker = trues(size(img_ori))
+    h, w = size(img_ori)
+    local img = FalsePadded(img_ori)
+    for i=1:h        
+        for j=1:w
+            if !img[i,j]
+                continue
+            end
+            p1 = img[i-1,j-1]
+            p2 = img[i-1,j]
+            p3 = img[i-1,j+1]
+            p4 = img[i  ,j+1]
+            p5 = img[i+1,j+1]
+            p6 = img[i+1,j]
+            p7 = img[i+1,j-1]
+            p8 = img[i  ,j-1]
+            local A  = (!p2  && (p3 || p4)) +
+                (!p4  && (p5 || p6)) +
+                (!p6  && (p7 || p8)) +
+                (!p8  && (p1 || p2))
+
+            local B1 = (p1 || p2) + (p3 || p4) + (p5 || p6) + (p7 || p8)
+            local B2 = (p2 || p3) + (p4 || p5) + (p6 || p7) + (p8 || p1)
+            local B  = min(B1,B2)
+
+            local G3 = false
+            if (odd_iteration)
+                G3 = (p2 || p3 || (!p5)) && p4                
+            else
+                G3 = (p6 || p7 || (!p1)) && p8
+            end
+            if (A == 1) && ((2<= B)  && (B <=3)) && (!G3) 
+                marker[i,j] =false
+            end
+        end
+        
+    end
+    img_ori[:] =  img_ori[:].&(marker[:])
+end    
+doc"""
+```julia
+function thinning(img::AbstractArray{T}; max_iterations=Inf) where T<:Bool
+```
+
+Applies a binary blob thinning operation, to achieve a skeletization of the input image.
+The procedure carries on until max_iterations is achieved or when there are no changes between two consecutives iterations.
+
+The algorithm is described in:
+1. Guo, Z., & Hall, R. W. (1989). Parallel thinning with two-subiteration algorithms. Communications of the ACM, 32(3), 359-373.
+2. Lam, L., Lee, S. W., & Suen, C. Y. (1992). Thinning methodologies-a comprehensive survey. IEEE Transactions on pattern analysis and machine intelligence, 14(9), 869-885.
+"""
+function thinning(img::AbstractArray{T}; max_iterations::I=0) where T<:Bool where I<:Integer
```

Comment:
We need to generalize this function signature in the future, for now please change it to:

```julia
function thinning(img::AbstractArray{Bool}; maxiter::Integer)
end
```